CMakeLists.txt

```diff
@@ -164,6 +154,29 @@ if(CUDA_VERSION VERSION_GREATER_EQUAL "7.5")
   endif()
 endif()
 
+if(CUDA_VERSION_MAJOR LESS 11)
+  # for some reason this line is necessary to propagate the standard to nvcc
+  # On MSVC this is not necessary / nvcc does not recognize the flag for MSVC
+  if(NOT MSVC)
+      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
+  endif()
+  set(CMAKE_CXX_STANDARD 11)
+  set(CMAKE_CXX_STANDARD_REQUIRED ON)
+  set(CMAKE_CUDA_STANDARD 11)
+  set(CMAKE_CUDA_STANDARD_REQUIRED ON)
+else()
+  # CUDA SDK 11 has a version of CUB that requires C++14
+  if(NOT MSVC)
+      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14")
+      list(APPEND CUDA_NVCC_FLAGS "-std=c++14")
+  endif()
+  set(CMAKE_CXX_STANDARD 14)
+  set(CMAKE_CXX_STANDARD_REQUIRED ON)
+  set(CMAKE_CUDA_STANDARD 14)
+  set(CMAKE_CUDA_STANDARD_REQUIRED ON)
+endif()
+
```

Comment:
Do you have an example of warnings that are generated? Are they so bothersome? It's a bit strange to have different standard for different versions of the sdk?

In any case it would be better to refactor the code:
```suggestion
set(PopSift_CXX_STANDARD 11)
if(CUDA_VERSION_MAJOR GREATER 10)
  set(PopSift_CXX_STANDARD 14)
endif()

# for some reason this line is necessary to propagate the standard to nvcc
# On MSVC this is not necessary / nvcc does not recognize the flag for MSVC
if(NOT MSVC)
    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++${PopSift_CXX_STANDARD}")
endif()
set(CMAKE_CXX_STANDARD ${PopSift_CXX_STANDARD})
set(CMAKE_CXX_STANDARD_REQUIRED ON)
set(CMAKE_CUDA_STANDARD ${PopSift_CXX_STANDARD})
set(CMAKE_CUDA_STANDARD_REQUIRED ON)
```
What do u think?

Reply:
I had been thinking about the shorter version that uses a macro for 11 or 14. I suppose that would be a good thing. The if(NOT MSVC) block would needed another line for adding -std=c++${PopSift_CXX_STANDARD} to CUDA_NVCC_FLAGS. Without it, Travis breaks in the Ubuntu 16.04 / SDK 11 combination.

PopSift uses Thrust only in s_filtergrid.cu, so warnings appear only when that file is compiled. The warning appears 3 times for every CC that you is configured in PopSift_CUDA_CC_LIST and this is one of the warnings:
```
In file included from /usr/local/cuda-11.0/include/thrust/detail/config/config.h:27:0,
                 from /usr/local/cuda-11.0/include/thrust/detail/config.h:23,
                 from /usr/local/cuda-11.0/include/thrust/copy.h:24,
                 from /home/griff/GIT/popsift/src/popsift/s_filtergrid.cu:21:
/usr/local/cuda-11.0/include/thrust/detail/config/cpp_dialect.h:104:13: warning: Thrust requires C++14. Please pass -std=c++14 to your compiler. Define THRUST_IGNORE_DEPRECATED_CPP_DIALECT to suppress this message.
   THRUST_COMPILER_DEPRECATION(C++14, pass -std=c++14 to your compiler);
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/local/cuda-11.0/include/cub/util_arch.cuh:36:0,
                 from /usr/local/cuda-11.0/include/thrust/system/cuda/detail/util.h:32,
                 from /usr/local/cuda-11.0/include/thrust/system/cuda/detail/for_each.h:34,
                 from /usr/local/cuda-11.0/include/thrust/system/detail/adl/for_each.h:42,
                 from /usr/local/cuda-11.0/include/thrust/detail/for_each.inl:27,
                 from /usr/local/cuda-11.0/include/thrust/for_each.h:279,
                 from /usr/local/cuda-11.0/include/thrust/system/detail/generic/transform.inl:19,
                 from /usr/local/cuda-11.0/include/thrust/system/detail/generic/transform.h:105,
                 from /usr/local/cuda-11.0/include/thrust/detail/transform.inl:25,
                 from /usr/local/cuda-11.0/include/thrust/transform.h:724,
                 from /usr/local/cuda-11.0/include/thrust/system/detail/generic/copy.inl:23,
                 from /usr/local/cuda-11.0/include/thrust/system/detail/generic/copy.h:58,
                 from /usr/local/cuda-11.0/include/thrust/detail/copy.inl:21,
                 from /usr/local/cuda-11.0/include/thrust/detail/copy.h:90,
                 from /usr/local/cuda-11.0/include/thrust/copy.h:511,
                 from /home/griff/GIT/popsift/src/popsift/s_filtergrid.cu:21:
/usr/local/cuda-11.0/include/cub/util_cpp_dialect.cuh:129:13: warning: CUB requires C++14. Please pass -std=c++14 to your compiler. Define CUB_IGNORE_DEPRECATED_CPP_DIALECT to suppress this message.
   CUB_COMPILER_DEPRECATION(C++14, pass -std=c++14 to your compiler);
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

In combination with parallel make, it would be hiding a lot of warnings.

Reply:
Perhaps instead of making this dependent on the CUDA SDK, we could require C++14 for everything. I'll commit that instead and see what Travis is saying.